### ISSUE TITLE
[apex] AST node attribute @Image not usable / always null in XPath rule / Designer

### ISSUE THREAD

**Issue by rsoesemann** [pmd/pmd#1568]
<img width="1160" alt="pmd_rule_designer__v_6_1_0_" src="https://user-images.githubusercontent.com/8180281/50896707-34d34400-140a-11e9-8599-49d41a7655a8.png">

**Affects PMD Version:** 
6.1

**Rule:**
Custom XPath Rule see screenshot

**Description:**
I am unable to see the contents of the Image AST Attribute and do regexp on it in XPath rules

**Code Sample demonstrating the issue:**

```
class MyApexClass {
    void bar(){
         if(!list.isEmpty() && list != null) {
            foo();
         }
    }
}
```

```
//*[matches(@Image, '&& [a-zA-Z0-9]* != null')]
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Designer from CLI

**Comment by oowekyala** [comment]
@rsoesemann `list` is a [reserved keyword](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_reserved_words.htm#!) in Apex. Just use an other variable name and the AST will be correctly parsed. 

First off you should not be trying to match the image of a very big node. The image is typically only populated on identifiers and such, so that expressions like you're writing will never be found in the image as a string. Instead, your XPath expression should match the AST structure. The AST looks like the following:
```
BooleanExpression // &&
+ PrefixExpression
  + MethodCallExpression
+ BooleanExpression // !=
  + VariableExpression
  + LiteralExpression
```

Now I agree there's a problem because the operators are not available as attributes, so that we can't differentiate the two BooleanExpressions. They should expose their operators somehow.

**Comment by rsoesemann** [comment]
I made up the code to make it more concise for the issue. It also doesn't work for valid code.

**Comment by rsoesemann** [comment]
I made up the code to make it more concise for the issue. It also doesn't work for valid code.

**Comment by oowekyala** [comment]
Sorry, I didn't mean to be rash. I proposed a fix in #1571, that solution doesn't populate the image but instead exposes an `@Op` attribute. The following XPath matches your node:
```ruby
//BooleanExpression[@Op="&&"][
  child::*[2][
    self::BooleanExpression[@Op="!="][
      child::*[1][self::VariableExpression] and 
      child::*[2][self::LiteralExpression[@LiteralType="NULL"]]
    ]
  ]
]
```

**Comment by oowekyala** [comment]
How did you overlay those sweet labels on your screenshot? :innocent:

**Comment by rsoesemann** [comment]
@oowekyala thank you so much for identifiying the issue so quickly and providing a fix (when will that be released?).

For screenshoting I use the free Skitch tool (https://evernote.com/intl/de/products/skitch)

**Comment by oowekyala** [comment]
@rsoesemann I hope it will be included in 6.11.0, which is due end of January.

Thanks for the link! Too bad it's Apple only

**Comment by adangel** [comment]
I've just merged the PR for 6.11.0 - the attributes will have slightly different names. The sample XPath expression from above will be:

```
//BooleanExpression[@Operator="&&"][
  child::*[2][
    self::BooleanExpression[@Operator="!="][
      child::*[1][self::VariableExpression] and 
      child::*[2][self::LiteralExpression[@LiteralType="NULL"]]
    ]
  ]
]
```

So, instead of "Op" the name is "Operator".

**Comment by rsoesemann** [comment]
![bildschirmfoto 2019-01-27 um 15 07 27](https://user-images.githubusercontent.com/8180281/51801985-48bdd900-2245-11e9-9a22-98a981163ac2.png)

What am I doing wrong because I see no match?!

**Comment by rsoesemann** [comment]
With this new extension would it be possible to detect the difference between those two method calls?

`System.debug('bad');`

and

`System.debug(LoggingLevel.ERROR, 'good');`

Finding such methods is easy: `//MethodCallExpression[lower-case(@FullMethodName)='system.debug']` but then the designers AST infos and my poor XPath skills come to an end...
